doc/flashproxy-reg-http.1
'\" t
.\"     Title: flashproxy-reg-http
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 05/07/2014
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "FLASHPROXY\-REG\-HTTP" "1" "05/07/2014" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
flashproxy-reg-http \- Register with a facilitator using the HTTP method
.SH "SYNOPSIS"
.sp
\fBflashproxy\-reg\-http\fR [\fIOPTIONS\fR] [\fIREMOTE\fR][:\fIPORT\fR]
.SH "DESCRIPTION"
.sp
Register with a flash proxy facilitator using an HTTP POST\&. By default the remote address registered is ":9000"\&.
.SH "OPTIONS"
.PP
\fB\-4\fR
.RS 4
Name lookups use only IPv4\&.
.RE
.PP
\fB\-6\fR
.RS 4
Name lookups use only IPv6\&.
.RE
.PP
\fB\-f\fR, \fB\-\-facilitator\fR=\fIURL\fR
.RS 4
Register with the given facilitator (default "https://fp\-facilitator\&.org/")\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display help message and exit\&.
.RE
.PP
\fB\-\-transport\fR=\fITRANSPORT\fR
.RS 4
Registrations include the fact that you intend to use the given
\fITRANSPORT\fR
(default "websocket")\&.
.RE
.PP
\fB\-\-unsafe\-logging\fR
.RS 4
Don\(cqt scrub IP addresses from logs\&.
.RE
.SH "SEE ALSO"
.sp
\fBhttp://crypto\&.stanford\&.edu/flashproxy/\fR
.SH "BUGS"
.sp
Please report using \fBhttps://trac\&.torproject\&.org/projects/tor\fR\&.


ChangeLog
  o Fixed a bug in the browser proxy which caused it to stop accepting
    new connections once it had failed 5 previous connections.

  o Updated the Tor Browser detection for the Firefox 24.0 User-Agent
    string. Patch by Arlo Breault. Fixes bug 11290.

Changes in version 1.6
  o Allowed the --port-forwarding option to work when the remote port
    number is given as 0.

  o Fixed registration on Mac OS X when the REMOTE address had an empty
    host part. A specification of ":9000", for example, would try to
    register "[]:9000".

  o Fixed registration on Windows with flashproxy-reg-appspot and
    flashproxy-reg-email. The certificate pinning code used a Python
    NamedTemporaryFile, which is not reopenable on Windows.

Changes in version 1.5
  o Add manpages for the facilitator and nodejs proxy, automatically
    generated by help2man.

  o Have nodejs flashproxy take GNU-style long command-line options.

  o Automate much of the configuration tasks involved in installing the
    facilitator using GNU autotools. See facilitator/INSTALL for details
    on the new process. Also move some common code here into the common
    python module. Patch by Ximin Luo. Fixes bug 9974.

  o Move common code to a separate flashproxy-common python module. Also
    split out some build scripts so distro packagers have an easier
    time. Patch by Ximin Luo. Fixes bug 6810.

  o Enabled binary frames (avoiding the overhead of base64 encoding) for
    Firefox 11 and later. Patch by Arlo Breault. Fixes bug 9069.

  o Removed a Python 2.7–dependent reference in flashproxy-reg-appspot.

Changes in version 1.4
  o Allowed websocket-server to gracefully handle SIGTERM.

  o Makefiles that install now obey DESTDIR to install relative to a
    different root.

  o Added a new observed Google public key pin for flashproxy-reg-email.

  o New --transport options in the client programs allow you to inform
    the facilitator that you want to receive connections of a certain
    kind. Transports other than the default "websocket" are
    experimental. Patch by George Kadianakis and David Fifield. Part of
    bug 9349.

  o Proxies now send a list of transport protocols they support
    (currently only "websocket"). This will allow the facilitator to
    assign proxies to clients that use matching transports. Patch by
    George Kadianakis. Part of bug 9349.

  o Allowed the facilitator to handle layered transports. For example, a
    client that register with the transport "obfs3|websocket" will
    receive a connection from a proxy using websocket, and will be
    connected to a relay that has an obfs3 server behind a websocket
    front end. Patch by Ximin Luo and George Kadianakis. Fixes bug 9349.

  o Changed to use the pluggable transport method name "flashproxy"
    rather than "websocket". Both names are equivalent and "websocket"
    continues to work. The reason for this change is to reduce confusion
    with a transport that simply makes a WebSocket connection to a
    "websocket" bridge, without receiving an inbound connection from a
    flash proxy. The default argument to the --transport option
    continues to be "websocket", because that option controls which
    particular protocol flash proxies should use to connect to you, and
    is distinct from the transport method name used by Tor.

  o Rearranged some files in the source tree. Facilitator documentation
    is now under facilitator/doc. The App Engine source code is under
    facilitator/appengine. The directory containing other ways to use
    the proxy moved from modules to proxy/modules. Patch by Ximin Luo.
    Fixes bug 9668.

Changes in version 1.3
  o Added a new observed Google public key pin.

Changes in version 1.2
  o The facilitator daemons have a --privdrop-user option that causes
    them to change to another user ID after reading keys and opening log
    files. facilitator-howto.txt shows how to configure them to use an
    unprivileged facilitator-nobody user. Patch by Alexandre Allaire and
    David Fifield. Fixes bug 8424.

  o Proxies now send the list of clients they are currently serving in
    their facilitator polling requests. This is meant to enable the
    facilitator to estimate the level of service each client is getting.
    Proxies send a protocol revision number "r=1" to signify the change.

  o The managed transport method name "flashproxy" is now recognized as
    a synonym for "websocket".

  o The badge localization now understands language subtags such as
    "ru-RU". Fixes bug 8828.

  o Language tags for badge localization are now case-insensitive.
    Patch by Eduardo Stalinho. Fixes bug 8829.

  o The badge localization is taken from the JavaScript property
    window.navigator.language when possible. Patch by Arlo Breault.
    Fixes bug 8827.

  o Proxies now attempt to connect to the client first, and only connect
    to the relay after the client connection is successful. This is
    meant to reduce the number of connections to the relay when clients
    haven't set up port forwarding. Introduced bug 9009, later fixed. 

  o A proxy no longer contacts the facilitator when it is given the
    "client" and "relay" parameters. It serves the one given client and
    then stops. Patch by Arlo Breault. Fixes bug 9006.

  o facilitator-email-poller ignores messages received a long time ago.
    This is to fix the situation where facilitator-email-poller stops
    running for some reason, comes back after some hours, and then
    flushes a lot of no-longer-relevant registrations out to proxies.
    Patch by Sukhbir Singh and David Fifield. Fixes bug 8285.

  o New --port-forwarding and friends options enable flashproxy-client
    to invoke tor-fw-helper to forward ports automatically. Patch by
    Arlo Breault and David Fifield. Fixes bug 9033.

  o The flash proxy, in debug mode, now hides potentially sensistive
    information like IP addresses. Patch by Arlo Breault. Fixes bug
    9170.


flashproxy-reg-http.py
#!/usr/bin/env python

import getopt
import httplib
import socket
import re
import sys
import urllib

DEFAULT_REMOTE_ADDR = ""
DEFAULT_REMOTE_PORT = 9000
DEFAULT_FACILITATOR_HOST = "tor-facilitator.bamsoftware.com"
DEFAULT_FACILITATOR_PORT = 9002

class options(object):
    facilitator_addr = None
    remote_addr = None

def usage(f = sys.stdout):
    print >> f, """\
Usage: %(progname)s [HOSTNAME][:PORT]
Register with a flash proxy facilitator using an HTTP POST. By default the
facilitator address is "%(fac_addr)s".

  -a, --address=ADDRESS  register the given address instead of \"%(remote_addr)s\".
  -h, --help             show this help. \
""" % {
    "progname": sys.argv[0],
    "fac_addr": format_addr((DEFAULT_FACILITATOR_HOST, DEFAULT_FACILITATOR_PORT)),
    "remote_addr": format_addr((DEFAULT_REMOTE_ADDR, DEFAULT_REMOTE_PORT)),
}

def parse_addr_spec(spec, defhost = None, defport = None):
    host = None
    port = None
    m = None
    # IPv6 syntax.
    if not m:
        m = re.match(ur'^\[(.+)\]:(\d+)$', spec)
        if m:
            host, port = m.groups()
            af = socket.AF_INET6
    if not m:
        m = re.match(ur'^\[(.+)\]:?$', spec)
        if m:
            host, = m.groups()
            af = socket.AF_INET6
    # IPv4 syntax.
    if not m:
        m = re.match(ur'^(.+):(\d+)$', spec)
        if m:
            host, port = m.groups()
            af = socket.AF_INET
    if not m:
        m = re.match(ur'^:?(\d+)$', spec)
        if m:
            port, = m.groups()
            af = 0
    if not m:
        host = spec
        af = 0
    host = host or defhost
    port = port or defport
    if not port:
        raise ValueError("Bad address specification \"%s\"" % spec)
    return host, int(port)

def format_addr(addr):
    host, port = addr
    if not host:
        return u":%d" % port
    # Numeric IPv6 address?
    try:
        addrs = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM, socket.IPPROTO_TCP, socket.AI_NUMERICHOST)
        af = addrs[0][0]
    except socket.gaierror, e:
        af = 0
    if af == socket.AF_INET6:
        return u"[%s]:%d" % (host, port)
    else:
        return u"%s:%d" % (host, port)

options.facilitator_addr = (DEFAULT_FACILITATOR_HOST, DEFAULT_FACILITATOR_PORT)
options.remote_addr = (DEFAULT_REMOTE_ADDR, DEFAULT_REMOTE_PORT)

opts, args = getopt.gnu_getopt(sys.argv[1:], "a:h", ["address=", "help"])
for o, a in opts:
    if o == "-a" or o == "--address":
        options.remote_addr = parse_addr_spec(a, DEFAULT_REMOTE_ADDR, DEFAULT_REMOTE_PORT)
    elif o == "-h" or o == "--help":
        usage()
        sys.exit()

if len(args) == 0:
    pass
elif len(args) == 1:
    options.facilitator_addr = parse_addr_spec(args[0], DEFAULT_FACILITATOR_HOST, DEFAULT_FACILITATOR_PORT)
else:
    usage(sys.stderr)
    sys.exit(1)

spec = format_addr(options.remote_addr)
http = httplib.HTTPConnection(*options.facilitator_addr)
try:
    http.request("POST", "/", urllib.urlencode({"client": spec}))
except Exception, e:
    print >> sys.stderr, "Failed to register: %s" % str(e)
    sys.exit(1)
http.close()

print "Registered \"%s\" with %s." % (spec, format_addr(options.facilitator_addr))
